go/vt/wrangler/reparent.go

```diff
@@ -425,7 +426,11 @@ func (wr *Wrangler) plannedReparentShardLocked(ctx context.Context, ev *events.R
 	wr.logger.Infof("promote slave %v", masterElectTabletAliasStr)
 	event.DispatchUpdate(ev, "promoting slave")
 	rp, err = wr.tmc.PromoteSlaveWhenCaughtUp(ctx, masterElectTabletInfo.Tablet, rp)
-	if err != nil {
+	if err != nil || (ctx.Err() != nil && ctx.Err() == context.DeadlineExceeded) {
+		// if this fails it is not enough to return an error. we should rollback all the changes made by DemoteMaster
+		if err1 := wr.tmc.UndoDemoteMaster(ctx, oldMasterTabletInfo.Tablet); err1 != nil {
```

Comment:
This will not work if the context has already expired. Is there a higher level context with longer timeout that we can use for this?

There should be one with the action timeout. Maybe use that, but with a shorter timeout.

Reply:
Ooh, I missed that. No, there is no higher level context, this is it.